ServiceManagerBase implementation + minor change to IServiceManager interface

samlaf opened the pull request:
We add a ServiceManagerBase abstract class that implements all of the "trivial" functions that just proxy to the slasher, and only leave for implementation:
- latestServeUntilBlock
- freezeOperator
- taskNumber

This way, AVS teams can derive their service manager from this base implementation without having to write a lot of redundant proxy contract calls.

This ServiceManagerBase depends on our registry contracts (registryCoordinator and stakeRegistry) so any team that uses their own registry will still need to implement their service manager from scratch. We could have an even more barebone ServiceManagerBase contract, but not sure how useful it would be.


There's one more change included in this PR, which is that I removed the `latestServeUntil` parameter from all the recordStakeUpdate calls in the IServiceManager interface, since all call sites were doing something like
```
uint32 latestServeUntilBlock = serviceManager.latestServeUntilBlock();
serviceManager.recordLastStakeUpdateAndRevokeSlashingAbility(operator, latestServeUntilBlock);
```
and I figured instead of doing these multiple separate calls to serviceManager, might as well instead of the serviceManager call its own `latestServeUntilBlock()` implementation. I left `latestServeUntilBlock` in the interface to make it explicit that the servicemanager should implement this public function, but I think it only makes sense to be called internally, so wondering whether to remove it entirely from the interface.

ChaoticWalrus commented on src/contracts/middleware/ServiceManagerBase.sol:
```diff
@@ -0,0 +1,145 @@
+
+    // PROXY CALLS TO EQUIVALENT SLASHER FUNCTIONS
+
+    /**
+     * @notice Called by the Registry in the event of a new registration, to forward a call to the Slasher
+     * @param operator The operator whose stake is being updated
+     */
+    function recordFirstStakeUpdate(
+        address operator
+    ) external onlyRegistryCoordinator {
+        slasher.recordFirstStakeUpdate(operator, latestServeUntilBlock());
+    }
+
+    /**
+     * @notice Called by the registryCoordinator, in order to forward a call to the Slasher, informing it of a stake update
+     * @param operator The operator whose stake is being updated
+     * @param updateBlock The block at which the update is being made
+     * @param prevElement The value of the previous element in the linked list of stake updates (generated offchain)
+     */
+    function recordStakeUpdate(
+        address operator,
+        uint32 updateBlock,
+        uint256 prevElement
+    ) external onlyRegistryCoordinatorOrStakeRegistry {
+        slasher.recordStakeUpdate(
+            operator,
+            updateBlock,
+            latestServeUntilBlock(),
+            prevElement
+        );
+    }
```
we might want to mark these functions as 'virtual' (so they can be overridden) and/or add extensible "hooks" to them, but this seems like a fine start

samlaf replied:
Made them virtual.

ChaoticWalrus commented:
Seems fine to me, in general.
Looks like there's a couple merge conflicts which need to be sorted out.
Requesting that @gpsanant take a quick look and resolve these merge conflicts.

gpsanant commented:
I don't think this change makes sense for AVS devs. I see the service manager as a proxy for the AVS to talk to the slasher and other EigenLayer contracts. What if we want more inputs into the `latestServeUntilBlock` calculation?

samlaf commented:
> I don't think this change makes sense for AVS devs. I see the service manager as a proxy for the AVS to talk to the slasher and other EigenLayer contracts. What if we want more inputs into the `latestServeUntilBlock` calculation?

It's a Base contract, so they don't *have* to use it. It's just meant as an example for teams to use. We're using it for Incredible-Squaring for eg. We can generalize it or add different ServiceManager base/template contracts like this as we expand and get a broader understanding of the types of AVS that are building on AVS.

samlaf commented:
@ChaoticWalrus @gpsanant reverted the IServiceManager changes to keep the same interface. Just removed the word "permissioned" in the comments since the interface doesn't enforce that permission, but can add it back if you guys think its better to keep it.

Also changed the merging branch from multiquorums to avs-unstable.

ChaoticWalrus commented:
> @ChaoticWalrus @gpsanant reverted the IServiceManager changes to keep the same interface. Just removed the word "permissioned" in the comments since the interface doesn't enforce that permission, but can add it back if you guys think its better to keep it.
> 
> Also changed the merging branch from multiquorums to avs-unstable.

My bad, I missed this update -- looks good to me, merge if you want.
